File: src/main/java/pw/switchcraft/plethora/integration/computercraft/registry/ComputerCraftMethodRegistration.java
package pw.switchcraft.plethora.integration.computercraft.registry;

import dan200.computercraft.api.ComputerCraftAPI;
import pw.switchcraft.plethora.api.method.IMethod;
import pw.switchcraft.plethora.api.method.IMethodRegistry;
import pw.switchcraft.plethora.api.module.IModuleContainer;
import pw.switchcraft.plethora.integration.computercraft.method.TurtleKineticMethods;

public final class ComputerCraftMethodRegistration {
    public static void registerMethods(IMethodRegistry r) {
        // Modules
        moduleMethod(r, "kinetic:use", TurtleKineticMethods.USE);
        moduleMethod(r, "kinetic:swing", TurtleKineticMethods.SWING);
    }

    private static <T> void method(IMethodRegistry r, String name, Class<T> target, IMethod<T> method) {
        r.registerMethod(ComputerCraftAPI.MOD_ID, name, target, method);
    }

    private static <T> void method(IMethodRegistry r, Class<T> target, IMethod<T> method) {
        r.registerMethod(ComputerCraftAPI.MOD_ID, method.getName(), target, method);
    }

    private static void moduleMethod(IMethodRegistry r, String name, IMethod<IModuleContainer> method) {
        method(r, name, IModuleContainer.class, method);
    }
}

File: src/main/java/io/sc3/plethora/integration/computercraft/registry/ComputerCraftMethodRegistration.kt
package io.sc3.plethora.integration.computercraft.registry

import dan200.computercraft.api.ComputerCraftAPI
import io.sc3.plethora.api.method.IMethod
import io.sc3.plethora.api.method.IMethodRegistry
import io.sc3.plethora.api.module.IModuleContainer
import io.sc3.plethora.integration.computercraft.InventoryMethodsWrapper
import io.sc3.plethora.integration.computercraft.method.TurtleKineticMethods
import net.minecraft.inventory.Inventory

object ComputerCraftMethodRegistration {
  @JvmStatic
  fun registerMethods(r: IMethodRegistry) {
    with (r) {
      // Modules
      moduleMethod("kinetic:use", TurtleKineticMethods.USE)
      moduleMethod("kinetic:swing", TurtleKineticMethods.SWING)

      // Inventory wrapper
      inventoryMethod("getSize", InventoryMethodsWrapper.GET_SIZE)
      inventoryMethod("list", InventoryMethodsWrapper.LIST)
      inventoryMethod("getItemDetail", InventoryMethodsWrapper.GET_ITEM_DETAIL)
      inventoryMethod("getItemLimit", InventoryMethodsWrapper.GET_ITEM_LIMIT)
      inventoryMethod("pushItems", InventoryMethodsWrapper.PUSH_ITEMS)
      inventoryMethod("pullItems", InventoryMethodsWrapper.PULL_ITEMS)
    }
  }

  private fun <T> IMethodRegistry.method(name: String, target: Class<T>, method: IMethod<T>) {
    registerMethod(ComputerCraftAPI.MOD_ID, name, target, method)
  }

  private fun IMethodRegistry.moduleMethod(name: String, method: IMethod<IModuleContainer>) {
    method(name, IModuleContainer::class.java, method)
  }

  private fun IMethodRegistry.inventoryMethod(name: String, method: IMethod<Inventory>) {
    method("inventory:$name", Inventory::class.java, method)
  }
}
